libraries/AP_Camera/AP_RunCam.cpp

```diff
@@ -0,0 +1,892 @@
+        } else {
+            // in the top-menu, keep track of the selected position
+            _top_menu_pos = (_top_menu_pos + 1) % RUNCAM_TOP_MENU_LENGTH;
+        }
+        break;
+
+    case Event::IN_MENU_EXIT:
+        // if we are in a sub-menu this will move us out, if we are in the root menu this will
+        // exit causing the state machine to get out of sync. the OSD menu hierachy is consistently
+        // 2 deep so we can count and be reasonably confident of where we are.
+        // the only exception is if someone hits save and exit on the root menu - then we are lost.
+        if (_in_menu-- > 0) {
+            simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_MODE);     // move up/out a menu
+            set_mode_change_timeout();
+        }
+        // no longer in the menu so trigger the OSD re-enablement
+        if (_in_menu <= 0) {
+            _state = State::EXITING_MENU;
+        }
+        break;
+
+    default:
+        break;
+    }
+}
+
+// enter the 2 key OSD menu
+void AP_RunCam::enter_2_key_osd_menu()
+{
+    // turn off built-in OSD so that the runcam OSD is visible
+    disable_osd();
+
+    simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_MODE);
+    set_button_press_timeout();
+    _in_menu = 1;
+    _state = State::IN_MENU;
+}
+
+// exit the 2 key OSD menu
+void AP_RunCam::exit_2_key_osd_menu()
+{
+    // turn built-in OSD back on
+    enable_osd();
+
+    if (_video_recording) {
+        simulate_camera_button(ControlOperation::RCDEVICE_PROTOCOL_CHANGE_START_RECORDING);
+        set_mode_change_timeout();
+    }
+    _state = State::READY;
+    _in_menu = 0;
+}
+
+// run the 5-key OSD simulation process
+void AP_RunCam::handle_5_key_simulation_process(Event ev) 
+{
+    switch (ev) {
+    case Event::BUTTON_RELEASE:
+        send_5_key_OSD_cable_simulation_event(ev);
+        _waiting_device_response = true;
+        return;
+        break;
+    case Event::ENTER_MENU:
+        if (_in_menu > 0) {
+            ev = Event::IN_MENU_RIGHT;
+        }
+        else {
+            // turn off built-in OSD so that the runcam OSD is visible
+            disable_osd();
+        }
+        break;
+
+    case Event::EXIT_MENU:
+        if (_in_menu > 0) {
+            // turn built-in OSD back on
+            enable_osd();
+        }
+        break;
+
+    default:
+        break;
+    }
+
+    if (ev != Event::NONE) {
+        send_5_key_OSD_cable_simulation_event(ev);
+        _is_button_pressed = true;
+        _waiting_device_response = true;
+    }
+}
+
+// handle a response
+void AP_RunCam::handle_5_key_simulation_response(const Response& response)
+{
+    if (response._result != ResponseStatus::SUCCESS) {
+        debug("response for command %d failed: %d\n", int(response._command), int(response._result));
+        simulation_OSD_cable_failed(response);
+        _waiting_device_response = false;
```

Comment:
Move this to the top of the function, remove this copy and the one at the end of the method.

Reply:
Fixed